### PR TITLE
spanconfigsql{watcher,reconciler}: setup SQLWatcher to watch for pts updates

### PR DESCRIPTION
**Pull request by adityamaru**
This change sets up a rangefeed on the `system.pts_records` table,
and adds this to the set of rangefeeds managed by the SQLWatcher. We
switch the unit emitted by the SQLWatcher to a union type called
`SQLUpdate` which captures either an update to a descriptor or to a
protected timestamp target. The zones and descriptor table rangefeeds
will continue to emit descriptor updates. The pts rangefeed will emit
descriptor updates for records that target schema objects, and pts target
updates for records that target cluster or tenants.

The SQLWatcher continues to dedup the SQLUpdates that it emits, so as to
ensure there is only one SQLUpdate per descriptor ID, and one SQLUpdate
per cluster or tenant target. The reconciler registers a handler that
is invoked everytime a batch of SQLUpdates are emitted by the SQLWatcher.
There is change in semantics in this part of the code, a future PR will teach
the reconciler to parse the pts target SQLUpdates, and in turn instruct the
SQLTranslator to generate the appropriate `SystemSpanConfigs`.

Note, this file moves the sqlwatcher tests into a ccl package so as to be
able to run backup statements to test the rangefeed on the pts table.

Informs: #73727

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
@irfansharif @arulajmani this is an early draft of what I had in mind for how the SQLWatcher would watch for protected timestamp updates. Before I clean up the tests and tidy up the PR I wanted inputs on some of the changes to the SQLWatcher interface.

**Comment by adityamaru** [comment]
This is RFAL now!

**Comment by adityamaru** [comment]
@irfansharif wrapped the target in a ProtectedTimestampUpdate - https://github.com/cockroachdb/cockroach/pull/75122/files#diff-879b43c0345e0967f0b1ee30422ea1993314a45a3195cb2cff030a62791ff434R309

Instead of storing it as separate Cluster and Tenant fields, I went with an opaque `ptpb.Target` field that we sanity check in the constructor. The reason I went with this is it becomes cleaner to dedup these entries in the watcher, based on the `ptpb.Target` stringer. Let me know what you think.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer.go:190]: Do we need to use a map here if the pts updates are sorted stably? Shouldn't we be able to compare against the last entry in the slice like we do above?

File: pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go:149]: [nit] Inline `require.NoError(t, buffer.Whatever)`, here and below.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:167]: I need the intermediate `prevUpdate` variable to be able to call `GetDescriptorUpdate` on it subsequently, and we only want to intialize prevUpdate if `i != 0`.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:153]: Hmm we could sort cluster before tenant, but if two tenant targets had the same timestamp I'm unsure how we would provide a stable sort. Unless we compare the proto Strings or something.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:190]: As mentioned above, I'm unsure how to make sorting between two TenantTargets stable. I think it is simpler to just use their string protos as a key as it is today?

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:126]: Done!

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:153]: Ended up sorting tenant updates before cluster, but the sort still isnt stable if two tenants or two cluster targets have the same timestamp.

File: pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go
Comment (adityamaru) [reply pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go:173]: Agreed, removed the count. Will rely on the reconciler tests in a follow up PR.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (irfansharif) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go]: Discussed a bit more offline, and also in relation to #74612 which this PR slightly overlaps with when sorting []rangefeedbuffer.Event. The proposal here was to represent an individual tenant protection as unit spanconfig.SQLUpdate, instead of representing it using an embedded slice, which would prevent sorting two elements of the type. Sorry for forgetting that we were using slices internally preventing the stable sort + necessitating the proto-string-for-map-key below.

**Comment by adityamaru** [comment]
@irfansharif changes to unwrap the Tenants_Target into individual SQLUpdates. Also made the sort stable, since each SQLUpdate only has one tenant ID in it now.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer.go:190]: [nit] s/events/bufferedEvents and s/bufferedEvents/events -- the latter is more precious.
[nit] s/type bufferedEvents/type events

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer.go:238]: [nit] Instead of this mutation, would it work to just iterate inline?

```go
for i, ev := range bufferedEvents[descriptorUpdatesIdx:] {
```

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer.go:254]: [nit] Can this be simplified?
```go
if prevUpdate.IsClusterUpdate() != curUpdate.IsClusterUpdate() {
```

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer.go:243]: Can we add a log.Fatal if prev event timestamp is > the current? Just being paranoid -- could do it when copying over elements 1-by-1.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (irfansharif) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:254]: Alternatively, if this case was at the last thing we were checking, we could just directly append to the slice.

```go
// If the previous buffered event, and the current event are both cluster
// target ProtectedTimestampUpdates we can ignore duplicates.
if prevUpdate.IsClusterUpdate() && curUpdate.IsClusterUpdate() {
  continue
}

// If the previous buffered event, and the current event are both tenant
// target ProtectedTimestampUpdates, compare tenantIDs and ignore
// duplicates.
if prevUpdate.IsTenantsUpdate() && curUpdate.IsTenantsUpdate() {
  if prevUpdate.TenantTarget != curUpdate.TenantTarget {
    sqlUpdates = append(sqlUpdates, update)
    continue
  }
}

// The previous buffered event and the current event are different
// target ProtectedTimestampUpdates, there is no deduplication possible.
sqlUpdates = append(sqlUpdates, update)
```

File: pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go:396]: [nit] For consistency with other similar logs:
```go
log.Infof(ctx, "established range feed over system.protected_ts_records starting at time %s", startTS)
```

File: pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go:379]: [nit] Add a default logcrash case here, like in the handlers above?

File: pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go:320]: [nit] Throw in a rand.Shuffle in here for good measure?

File: pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go:284]: [nit] Use linebreaks.

**Review by irfansharif — APPROVED** [review]
LGTM, nice work! Left only nits.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:243]: hmm adding it here might not be right cause we major sort on tenant ID, so you could have:

`event{ tenantTarget:1, ts:7}, event{tenantTarget:2, ts5}`

I added it when we are copying over yup.

**Review by arulajmani — APPROVED** [review]
I only have a few nits! Good stuff :lgtm: 

> Reviewed 1 of 13 files at r1, 10 of 12 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-:-MuwMwcPEg_c6prdb-x3:b-wnsjtn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @irfansharif)

___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwHqHj5mecrY4RmJrr:-MuwHqHj5mecrY4RmJrs:b-3jo5fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer.go#L109)):*
> ```Go
> 
> // Less implements the sort.Interface methods.
> func (s bufferedEvents) Less(i, j int) bool {
> ```

nit: do you think it's worth capturing the sort order here? I know you already have it inlined; your call. 
___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer.go, line 255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwM6IN9s3KEwJh8CsR:-MuwM6IN9s3KEwJh8CsS:brepkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer.go#L255)):*
> ```Go
> 		// target ProtectedTimestampUpdates there is no deduplication possible.
> 		if prevUpdate.IsClusterUpdate() && curUpdate.IsTenantsUpdate() ||
> 			prevUpdate.IsTenantsUpdate() && curUpdate.IsClusterUpdate() {
> ```

Can we ever have `prevUpdate.IsClusterUpdate()` and `curUpdate.IsTenantUpdate`?
___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer.go, line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwMYcs3vXbvOLfcP3g:-MuwMYcs3vXbvOLfcP3h:b-3bezh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer.go#L266)):*
> ```Go
> 		}
> 
> 		// If the previous buffered event, and the current event are both tenant
> ```

nit: move this at the top of the loop considering we sort tenant targets before cluster targets? 
___
*[pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go, line 374 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwLNZ8-XCDMRkjNtca:-MuwLNZ8-XCDMRkjNtcb:b-r61pwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go#L374)):*
> ```Go
> 		case *ptpb.Target_SchemaObjects:
> 			// For PTS records with schema object targets, unwrap the descriptor IDs,
> 			// and emit them as descriptor SQLUpdates. This allows for the deduplication
> ```

Coming back to this after reading your tests for the buffer, hijacking PTS updates on descriptor IDs at this level worked out well :100: 
___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwKXlMD2fGM8sXbQlS:-MuwKXlMD2fGM8sXbQlT:bkmngqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go#L109)):*
> ```Go
> 	buffer.advance(descriptorsRangefeed, ts(10))
> 	require.NoError(t, err)
> 	buffer.advance(protectedTimestampRangefeed, ts(10))
> ```

nit: here, and below, instead of moving descriptors/pts rangefeeds in lockstep you could make the test case more interesting by choosing different values and making sure the lowest of the 3 is the combined frontier timestamp.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:238]: Left it as is cause `prevUpdate` also indexes into the slice using `i-1` which seems more intuitive than `descriptorUpdatesIdx + i - 1`.

File: pkg/spanconfig/spanconfigsqlwatcher/buffer.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/buffer.go:190]: did s/events/bufferedEvents and s/type bufferedEvents/type events

but the actual slice we care about cant shadow the type name according to lint 🤷 so called it something other than `events`.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75122)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @arulajmani, and @irfansharif)

___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwHqHj5mecrY4RmJrr:-MuwXMVD90EJO323tdza:bvk8tab) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer.go#L109)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: do you think it's worth capturing the sort order here? I know you already have it inlined; your call. 
</blockquote></details>

Heh, I did have it here for a second and then moved it back inline cause it felt more natural to highlight to the reader that this sort.Sort is actually more involved than meets the eye.
___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer.go, line 255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwM6IN9s3KEwJh8CsR:-MuwXxYS4CJamY-5vcCx:b-xdw723) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer.go#L255)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we ever have `prevUpdate.IsClusterUpdate()` and `curUpdate.IsTenantUpdate`?
</blockquote></details>

I refactored this to Irfan's suggestion above. You're right I don't think we could have in the older version of the code.
___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer.go, line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwMYcs3vXbvOLfcP3g:-MuwY3ON7g4AknEqQiqj:biqldxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer.go#L266)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: move this at the top of the loop considering we sort tenant targets before cluster targets? 
</blockquote></details>

refactored to irfan's suggestion but also moveed this clause higher.
___
*[pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75122#-MuwKXlMD2fGM8sXbQlS:-MuwdzMH-Uw2rjooqjrV:bg44oou) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0ce909a211a8a7f635545f6138a78fee1bb7cc1/pkg/spanconfig/spanconfigsqlwatcher/buffer_test.go#L109)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: here, and below, instead of moving descriptors/pts rangefeeds in lockstep you could make the test case more interesting by choosing different values and making sure the lowest of the 3 is the combined frontier timestamp. 
</blockquote></details>

staggered the timestamps as suggested.

**Comment by adityamaru** [comment]
Thanks for all the review on this!

bors r=irfansharif,arulajmani

**Comment by otan** [comment]
```
[04:31:38][run] pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go:406:21: not enough arguments in call to rangefeed.WithDiff
[04:31:38][run] 	have ()
[04:31:38][run] 	want (bool)
```

merge skew :(

(yay bors!!!)

**Comment by adityamaru** [comment]
CI failure is https://github.com/cockroachdb/cockroach/issues/75960.

bors r+

**Pull request merged**
